src/OpenTelemetry.AutoInstrumentation.Native/continuous_profiler.cpp

```diff
@@ -0,0 +1,1076 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+// We want to use std::min, not the windows.h macro
+#define NOMINMAX
+#include "continuous_profiler.h"
+#include "logger.h"
+#include <chrono>
+#include <map>
+#include <algorithm>
+#include <shared_mutex>
+#ifndef _WIN32
+#include <pthread.h>
+#include <codecvt>
+#endif
+
+constexpr auto kMaxStringLength = 512UL;
+
+constexpr auto kMaxCodesPerBuffer = 10 * 1000;
+
+// If you change this, consider ThreadSampler.cs too
+constexpr auto kSamplesBufferMaximumSize = 200 * 1024;
+
+constexpr auto kSamplesBufferDefaultSize = 20 * 1024;
+
+// If you change these, change ThreadSampler.cs too
+constexpr auto kDefaultSamplePeriod = 10000;
+constexpr auto kMinimumSamplePeriod = 1000;
+
+constexpr auto kDefaultMaxAllocsPerMinute = 200;
+
+// FIXME make configurable (hidden)?
+// These numbers were chosen to keep total overhead under 1 MB of RAM in typical cases (name lengths being the biggest
+// variable)
+constexpr auto kMaxFunctionNameCacheSize         = 5000;
+constexpr auto kMaxVolatileFunctionNameCacheSize = 2000;
+
+// If you squint you can make out that the original bones of this came from sample code provided by the dotnet project:
+// https://github.com/dotnet/samples/blob/2cf486af936261b04a438ea44779cdc26c613f98/core/profiling/stacksampling/src/sampler.cpp
+// That stack sampling project is worth reading for a simpler (though higher overhead) take on thread sampling.
+
+/*
+  Locking/threading design:
+  We have the following shared data structures:
+  - A buffer for captured thread samples, used by a single writing thread and a single (managed) reading one
+  - A buffer for captured allocation samples, used by any application thread at any time
+  - a name cache (data structure for building humnan-readable stack traces), used during both thread and
+    allocation sampling
+  - a cache of thread id->thread name (set by each thread itself, used during either sample type)
+  - a cache of thread span context state (set by any application thread at any time, used during either sample type)
+
+  In general we want to keep locks "adjacent" to just one data structure and usage of them local to one
+  modifying/reading method to simplify analysis.  However, there are some special cases.
+  Here are the locks in use:
+  - cpu_buffer_lock guarding access to the thread samples buffer
+  - allocation_buffer_lock guarding access to the buffer for allocation samples
+  - name_cache_lock, guarding the
```

Comment:
This looks to be missing the rest of the line

Reply:
😞 Oops!  `guarding the data structures used for function/class name lookup`